src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    internal class MultipartSectionPipeReader : PipeReader
+    {
+        private readonly PipeReader _pipeReader;
+        private bool _finished = false;
+        private bool _metadataSkipped = false;
+        private readonly MultipartBoundary _boundary;
+        private int _partialMatchIndex = 0;
+        private ReadOnlySequence<byte> _buffer;
+        private SequencePosition _endPosition;
+        private bool _isReaderComplete;
+
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+        private static ReadOnlySpan<byte> EndOfFileDelimiter => new byte[] { (byte)'-', (byte)'-' };
+
+        public long RawLength { get; private set; } = 0;
+
+        public long? LengthLimit { get; private set; }
+        // public long Length { get; private set; } = 0;
+
+        internal MultipartSectionPipeReader(PipeReader pipeReader, MultipartBoundary boundary)
+        {
+            _pipeReader = pipeReader;
+            _boundary = boundary;
+        }
+
+        private void UpdateLength(long read)
+        {
+            //Length += read;
+            RawLength += read;
+            if (LengthLimit.HasValue && RawLength > LengthLimit.GetValueOrDefault())
+            {
+                throw new InvalidDataException($"Multipart body length limit {LengthLimit.GetValueOrDefault()} exceeded.");
+            }
+        }
+
+        public override Stream AsStream(bool leaveOpen = false)
+        {
+            return new PipeReaderStream(this, leaveOpen);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+
+            ThrowIfCompleted();
+
+            if (consumed.GetObject() == null || examined.GetObject() == null)
+            {
+                return;
+            }
+
+            if (!_buffer.IsEmpty)
+            {
+                var buffer = _buffer.Slice(consumed); //not sure how to handle examined since this is relevant only at the end of the pipe.
+                UpdateLength(_buffer.Length - buffer.Length);
```

Comment:
I don't think you need to track RawLength on Advance, only in ReadAsync.

Reply:
I think you are right

Reply:
Wait, the reason I put UpdateLength in the advance is because unless advance is called ReadAsync can read the same buffer over and over. I think it will be a lot more difficult, and not necessarily better to try and create a mechanism to avoid duplicate UpdateLength.